### PR TITLE
Feature/add functions

### PR DESCRIPTION
**Pull request by proind**
a few changes :

1.added a "create" static method to Function. Currently Function  only wraps existing functions in IDA, I needed a way to create new functions , and this seemed like a good solution. 
2. added type property to line: that way you can tell what type of line it is without going over all the types  by hand . also added the type to Line's repr .( the repr part felt convenient to me , but maybe unnecessary) 
3.added a type parameter to segments() function , so you get can a segment of a specific type, CODE segment for example

### CODE REVIEW

**Comment by tmr232** [comment]
Thanks for the contribution!
I'll go over the code and play with it soon.

File: sark/code/segment.py [hunk 1]
```diff
@@ -244,6 +247,11 @@ def __repr__(self):
                                         self.bitness)
 
 
-def segments():
+def segments(type=None):
```
Comment (tmr232) [inline sark/code/segment.py]: Please do not shadow builtins.

File: sark/code/line.py [hunk 2]
```diff
@@ -180,9 +180,19 @@ def endEA(self):
     def disasm(self):
         """Line Disassembly"""
         return idc.GetDisasm(self.ea)
+    @property
+    def type(self):
```
Comment (tmr232) [inline sark/code/line.py]: Is this meant for text-output only, or do you have other uses for this property?

**Comment by tmr232** [comment]
Just went over the code quickly now - looks good (except for minor comment), but I'll have to play with it for a bit.

File: sark/code/line.py (hunk 2 above)
Comment (proind) [inline sark/code/line.py]: yeah , text-output only

File: sark/code/function.py [hunk 3]
```diff
@@ -86,10 +86,21 @@ def __init__(self, ea=UseCurrentAddress, name=None):
 
         elif isinstance(ea, Line):
             ea = ea.ea
-
         self._func = get_func(ea)
         self._comments = Comments(self)
 
+    @staticmethod
+    def create(ea=UseCurrentAddress, name=None):
```
Comment (tmr232) [inline sark/code/function.py]: What is the `name` parameter for? It does not seem to be used.

File: sark/code/function.py [hunk 4]
```diff
@@ -86,10 +86,21 @@ def __init__(self, ea=UseCurrentAddress, name=None):
...
+        try:
+            Function(ea,name)
+            raise exceptions.SarkFunctionExists("function already exists")
+        except exceptions.SarkNoFunction:
```
Comment (tmr232) [inline sark/code/function.py]: I think we should an `is_function(ea)` helper, as exceptions should deal with exceptional cases, and "no function found" is the common case here.

File: sark/code/segment.py [hunk 5]
```diff
@@ -244,6 +247,11 @@ def __repr__(self):
                                         self.bitness)
 
 
-def segments():
+def segments(seg_type=None):
+    ''' :param seg_type: type of segment e.g. SEG_CODE
+        :return: segment based on type. if type is None, returns all segments
```
Comment (tmr232) [inline sark/code/segment.py]: I try to follow this - https://pypi.python.org/pypi/sphinxcontrib-napoleon - in the docstrings, as it is more readable for devs, and gets parse just as well.

File: sark/code/line.py [hunk 6]
```diff
@@ -180,9 +180,19 @@ def endEA(self):
...
+                      self.is_data:"data",
+                      self.is_string:"string",
+                      self.is_tail:"tail",
+                      self.is_unknown:"unknown"}
+        for k, v in properties.items():
+            if k: return v
 
     def __repr__(self):
-        return "[{:08X}]    {}".format(self.ea, self.disasm)
+        return "[{:08X}]    {} {}".format(self.ea, self.disasm, self.type)
```
Comment (tmr232) [inline sark/code/line.py]: Is this beneficial? Isn't it obvious when looking at the line?

File: sark/code/function.py (hunk 3 above)
Comment (proind) [inline sark/code/function.py]: you are right , removed it

File: sark/code/function.py (hunk 4 above)
Comment (proind) [inline sark/code/function.py]: let me know if the change I made is what you had in mind

File: sark/code/line.py (hunk 6 above)
Comment (proind) [inline sark/code/line.py]: I added it when debugging a script that dealt with data ,and ignored unknowns. wanted to be able to tell what is what at a glance 
for example in cases like this 
`DCB 0x33 ; 3 `
this was marked as unknown , I'd assume it's data 
if you think it's unnecessary in most cases I can remove it

File: sark/code/function.py [hunk 7]
```diff
@@ -86,10 +86,30 @@ def __init__(self, ea=UseCurrentAddress, name=None):
...
+            return False
+
+    @staticmethod
+    def create(ea=UseCurrentAddress):
+        if ea == Function.UseCurrentAddress:
+            ea = idc.here()
+
+        if Function.is_function(ea):
+            raise exceptions.SarkFunctionExists("function already exists")
```
Comment (tmr232) [inline sark/code/function.py]: Nit: 'f' -> 'F'

File: sark/code/function.py [hunk 8]
```diff
@@ -86,10 +86,30 @@ def __init__(self, ea=UseCurrentAddress, name=None):
...
+        else:
```
Comment (tmr232) [inline sark/code/function.py]: No need for the `else` here.

File: sark/code/line.py (hunk 6 above)
Comment (tmr232) [inline sark/code/line.py]: If you are using a script, you should query the object, not parse text.
I think the `__repr__` is mainly for shell use, so the type is not relevant here.

Comment (proind) [inline sark/code/line.py]: I meant that I used it from the shell in order to understand why a script has issues.  
The script queried the object , but I wanted a convenient way to see the same data from the shell .

Comment (tmr232) [inline sark/code/line.py]: I think that in that case we can do without it. Sorry.

Comment (proind) [inline sark/code/line.py]: sure, removed .

**Comment by tmr232** [comment]
Merged 👍 

Checking your code, I also found an existing bug in Sark, where some lines were wrongfully detected as strings.
This is because `idc.GetStringType` can return either `-1` or `None` on failure. I fixed it.

**Pull request merged**